File: sdks/ts/src/revocation.ts
/**
 * Revocation cache + kill-switch handling (doc 11 §7, Ruling C11).
 *
 * gatekeeper's revocation-service emits `tasks.revocations.v1` (durable
 * JetStream); the Orchestrator maps those to `control.kill` (CORE NATS
 * broadcast, no stream). PEPs must ACK and halt in-flight work ≤ 5 s
 * (graceful: stop new requests, drain ≤ 5 s, report execution.halted).
 *
 * This cache is the local revocation set the PEP consults before every
 * network action; entries honor `expires_at` (temporary revocations lift).
 * Fail-safe direction: an unparseable control.kill broadcast is treated as a
 * GLOBAL kill — kill signals fail toward halting, never toward continuing.
 */

import { fromBinary } from "@bufbuild/protobuf";
import { anyUnpack, timestampDate } from "@bufbuild/protobuf/wkt";
import { EnvelopeSchema } from "@aegisbastion/gen/aegisbastion/platform/v1/bus_pb.js";
import {
  RevocationEventSchema,
  RevocationScope,
  type Revocation,
  type RevocationEvent,
} from "@aegisbastion/gen/aegisbastion/gatekeeper/v1/revocation_pb.js";
import { PepError } from "./errors.js";
import { canonicalizeTarget } from "./scope.js";
import type { ScopeTokenClaims } from "./token.js";

interface CacheEntry {
  revocationId: string;
  expiresAtMs: number | null;
}

export type KillSignal =
  | { kind: "global"; reason: string }
  | { kind: "roe"; roeId: string; reason: string }
  | { kind: "target"; target: string; reason: string }
  | { kind: "capability"; capability: string; reason: string };

export class RevocationCache {
  private global: CacheEntry | null = null;
  private readonly roes = new Map<string, CacheEntry>();
  private readonly targets = new Map<string, CacheEntry>();
  private readonly capabilities = new Map<string, CacheEntry>();
  private readonly seenRevocationIds = new Set<string>();

  constructor(private readonly nowMs: () => number = () => Date.now()) {}

  /** Apply one Revocation (idempotent on revocation_id). */
  apply(rev: Revocation): void {
    if (rev.revocationId && this.seenRevocationIds.has(rev.revocationId)) return;
    if (rev.revocationId) this.seenRevocationIds.add(rev.revocationId);
    const entry: CacheEntry = {
      revocationId: rev.revocationId,
      expiresAtMs: rev.expiresAt ? timestampDate(rev.expiresAt).getTime() : null,
    };
    switch (rev.scope) {
      case RevocationScope.GLOBAL:
        this.global = entry;
        break;
      case RevocationScope.ROE:
        if (rev.key) this.roes.set(rev.key, entry);
        break;
      case RevocationScope.TARGET: {
        if (rev.key) {
          let canonical = rev.key;
          try {
            canonical = canonicalizeTarget(rev.key).canonical;
          } catch {
            // Unparseable target key: store raw — matching is done on
            // canonicalized targets, so a raw entry can only under-match,
            // and the GLOBAL/ROE scopes still cover the halt.
          }
          this.targets.set(canonical, entry);
        }
        break;
      }
      case RevocationScope.CAPABILITY:
        if (rev.key) this.capabilities.set(rev.key, entry);
        break;
      default:
        // Unknown scope: fail-safe — treat as global halt.
        this.global = entry;
        break;
    }
  }

  /** Apply a bus RevocationEvent. */
  applyEvent(event: RevocationEvent): void {
    if (event.revocation) this.apply(event.revocation);
  }

  private live(entry: CacheEntry | null | undefined): entry is CacheEntry {
    return entry != null && (entry.expiresAtMs === null || entry.expiresAtMs > this.nowMs());
  }

  /**
   * The kill signal applying to this token/target/capability right now, or
   * null when nothing is revoked. Checked before EVERY network action.
   */
  check(claims: ScopeTokenClaims, rawTarget?: string, capability?: string): KillSignal | null {
    if (this.live(this.global)) {
      return { kind: "global", reason: "platform-wide revocation active" };
    }
    if (this.live(this.roes.get(claims.roe_id))) {
      return { kind: "roe", roeId: claims.roe_id, reason: `RoE ${claims.roe_id} revoked` };
    }
    if (capability !== undefined && this.live(this.capabilities.get(capability))) {
      return { kind: "capability", capability, reason: `capability ${capability} revoked` };
    }
    for (const cap of claims.capabilities) {
      if (this.live(this.capabilities.get(cap))) {
        return { kind: "capability", capability: cap, reason: `capability ${cap} revoked` };
      }
    }
    if (rawTarget !== undefined) {
      let canonical: string | null = null;
      try {
        canonical = canonicalizeTarget(rawTarget).canonical;
      } catch {
        canonical = null;
      }
      const entry =
        (canonical !== null ? this.targets.get(canonical) : undefined) ?? this.targets.get(rawTarget);
      if (this.live(entry)) {
        return { kind: "target", target: rawTarget, reason: `target revoked` };
      }
    }
    return null;
  }

  /** Throw PepError(REVOKED) when any revocation applies. */
  assertNotRevoked(claims: ScopeTokenClaims, rawTarget?: string, capability?: string): void {
    const signal = this.check(claims, rawTarget, capability);
    if (signal !== null) {
      throw new PepError("REVOKED", signal.reason, { kind: signal.kind });
    }
  }

  get size(): number {
    return (
      (this.global ? 1 : 0) + this.roes.size + this.targets.size + this.capabilities.size
    );
  }
}

/**
 * Decode a `control.kill` broadcast payload (CORE NATS, no JetStream stream,
 * doc 01 §8.1). Contract form: an Envelope whose Any payload is a gatekeeper
 * RevocationEvent (Ruling C11: the Orchestrator maps revocations to
 * control.kill). Fail-SAFE: anything unparseable is treated as a GLOBAL kill.
 */
export function decodeControlKill(data: Uint8Array): RevocationEvent | { global: true } {
  try {
    const envelope = fromBinary(EnvelopeSchema, data);
    if (envelope.payload) {
      const event = anyUnpack(envelope.payload, RevocationEventSchema);
      if (event) return event;
    }
  } catch {
    // fall through to fail-safe global kill
  }
  return { global: true };
}

File: sdks/go/pep/revocation.go
package pep

import (
	"sync"
	"time"

	gatekeeperv1 "github.com/aegisbastion/aegisbastion/gen/go/aegisbastion/gatekeeper/v1"
	platformv1 "github.com/aegisbastion/aegisbastion/gen/go/aegisbastion/platform/v1"

	"github.com/aegisbastion/aegisbastion/sdks/go/scope"
)

// RevocationCache is the PEP-side revocation set (doc 11 §2.1.7/§7): fed by
// tasks.revocations.v1 events (≤ 5 s halt SLA) and reconciled against
// RevocationService.ListRevocations at ≤ 30 s. It answers, for every intended
// action, "has this been revoked?" — global, per-RoE, per-target, or
// per-capability.
//
// Fail-closed posture: a missed refresh keeps the last known set (never
// silently clears); a revoked entry without expiry holds until a snapshot
// from the service lifts it.
type RevocationCache struct {
	now func() time.Time

	mu      sync.RWMutex
	entries map[revocationKey]revocationEntry
}

type revocationKey struct {
	scope gatekeeperv1.RevocationScope
	key   string // canonicalized for TARGET scope
}

type revocationEntry struct {
	id        string
	expiresAt time.Time // zero = until lifted
	reason    string
}

// RevocationCacheOption configures a RevocationCache.
type RevocationCacheOption func(*RevocationCache)

// WithRevocationClock injects a clock (tests).
func WithRevocationClock(now func() time.Time) RevocationCacheOption {
	return func(c *RevocationCache) { c.now = now }
}

// NewRevocationCache builds an empty cache.
func NewRevocationCache(opts ...RevocationCacheOption) *RevocationCache {
	c := &RevocationCache{
		now:     time.Now,
		entries: map[revocationKey]revocationEntry{},
	}
	for _, o := range opts {
		o(c)
	}
	return c
}

func canonicalRevKey(s gatekeeperv1.RevocationScope, key string) revocationKey {
	if s == gatekeeperv1.RevocationScope_REVOCATION_SCOPE_TARGET {
		if t, err := scope.Canonicalize(key); err == nil {
			key = t.Canonical
		}
	}
	return revocationKey{scope: s, key: key}
}

// ApplyEvent applies one tasks.revocations.v1 event (the ≤ 5 s halt path).
func (c *RevocationCache) ApplyEvent(evt *gatekeeperv1.RevocationEvent) {
	if evt == nil || evt.GetRevocation() == nil {
		return
	}
	c.Apply(evt.GetRevocation())
}

// Apply records one revocation. Expired temporary revocations are dropped.
func (c *RevocationCache) Apply(rev *gatekeeperv1.Revocation) {
	if rev == nil || rev.GetScope() == gatekeeperv1.RevocationScope_REVOCATION_SCOPE_UNSPECIFIED {
		return
	}
	var exp time.Time
	if rev.GetExpiresAt() != nil {
		exp = rev.GetExpiresAt().AsTime()
		if !exp.IsZero() && !c.now().Before(exp) {
			return // already lapsed
		}
	}
	k := canonicalRevKey(rev.GetScope(), rev.GetKey())
	c.mu.Lock()
	c.entries[k] = revocationEntry{id: rev.GetRevocationId(), expiresAt: exp, reason: rev.GetReason()}
	c.mu.Unlock()
}

// Replace swaps in a full snapshot from RevocationService.ListRevocations —
// the ≤ 30 s reconciliation that lifts cleared revocations and backfills
// missed events.
func (c *RevocationCache) Replace(revs []*gatekeeperv1.Revocation) {
	fresh := map[revocationKey]revocationEntry{}
	for _, rev := range revs {
		if rev == nil || rev.GetScope() == gatekeeperv1.RevocationScope_REVOCATION_SCOPE_UNSPECIFIED {
			continue
		}
		var exp time.Time
		if rev.GetExpiresAt() != nil {
			exp = rev.GetExpiresAt().AsTime()
			if !exp.IsZero() && !c.now().Before(exp) {
				continue
			}
		}
		fresh[canonicalRevKey(rev.GetScope(), rev.GetKey())] = revocationEntry{
			id: rev.GetRevocationId(), expiresAt: exp, reason: rev.GetReason(),
		}
	}
	c.mu.Lock()
	c.entries = fresh
	c.mu.Unlock()
}

// Revoked reports whether an action under (roeID, capability, target) is
// revoked, and why. GLOBAL revokes everything. target may be "" (then the
// TARGET scope is not consulted).
func (c *RevocationCache) Revoked(roeID, capability, target string) (bool, string) {
	c.mu.RLock()
	defer c.mu.RUnlock()
	now := c.now()
	hit := func(k revocationKey) (string, bool) {
		e, ok := c.entries[k]
		if !ok {
			return "", false
		}
		if !e.expiresAt.IsZero() && !now.Before(e.expiresAt) {
			return "", false
		}
		reason := e.reason
		if reason == "" {
			reason = "revocation " + e.id
		}
		return reason, true
	}
	if r, ok := hit(revocationKey{scope: gatekeeperv1.RevocationScope_REVOCATION_SCOPE_GLOBAL}); ok {
		return true, "global revocation: " + r
	}
	if roeID != "" {
		if r, ok := hit(revocationKey{scope: gatekeeperv1.RevocationScope_REVOCATION_SCOPE_ROE, key: roeID}); ok {
			return true, "RoE revocation: " + r
		}
	}
	if capability != "" {
		if r, ok := hit(revocationKey{scope: gatekeeperv1.RevocationScope_REVOCATION_SCOPE_CAPABILITY, key: capability}); ok {
			return true, "capability revocation: " + r
		}
	}
	if target != "" {
		if r, ok := hit(canonicalRevKey(gatekeeperv1.RevocationScope_REVOCATION_SCOPE_TARGET, target)); ok {
			return true, "target revocation: " + r
		}
	}
	return false, ""
}

// Halted reports whether a GLOBAL revocation is active.
func (c *RevocationCache) Halted() (bool, string) {
	return c.Revoked("", "", "")
}

// KillDecision interprets a control.kill broadcast envelope (core NATS, no
// stream) for this agent. The Orchestrator maps gatekeeper revocations onto
// control.kill (Ruling C11), so the payload is a gatekeeper RevocationEvent
// when typed; a per-mission kill may instead arrive as an envelope whose
// mission_id matches with no typed payload. Pure function — side-effect free
// so it is trivially testable.
//
// missionID/agentID scope the decision: a kill for another mission or agent
// does not halt this one.
func KillDecision(env *platformv1.Envelope, missionID string) (kill bool, reason string) {
	if env == nil {
		return false, ""
	}
	if env.GetPayload() != nil {
		if evt, err := env.GetPayload().UnmarshalNew(); err == nil {
			if rev, ok := evt.(*gatekeeperv1.RevocationEvent); ok {
				r := rev.GetRevocation()
				switch r.GetScope() {
				case gatekeeperv1.RevocationScope_REVOCATION_SCOPE_GLOBAL:
					return true, "global kill: " + r.GetReason()
				case gatekeeperv1.RevocationScope_REVOCATION_SCOPE_ROE,
					gatekeeperv1.RevocationScope_REVOCATION_SCOPE_TARGET,
					gatekeeperv1.RevocationScope_REVOCATION_SCOPE_CAPABILITY:
					// Scoped revocations halt only tasks matching the key —
					// the Guard consults the RevocationCache per target;
					// a broadcast of one is a platform-wide signal to check.
					return true, "scoped kill (" + r.GetScope().String() + " " + r.GetKey() + "): " + r.GetReason()
				}
			}
		}
	}
	// Untyped per-mission kill: envelope mission_id matches ours.
	if missionID != "" && env.GetMissionId() == missionID {
		return true, "mission kill: " + missionID
	}
	return false, ""
}
